Query timeout is not respected when host is not listening

`com.github.jasync-sql:jasync-postgresql:1.1.6`

Steps to reproduce:
* Set up a connection pool with connection & query timeout (100ms in my case)
* Stop the server the client is pointing at
* Run the provided code

Observed behavior: only 1 exception (this number depends on `maxActiveConnections`, e.g. if I set it to 5, I will likely get 5 exceptions in the output) is printed out and the rest of the coroutines freeze, until the server is back up (easy to test by stopping and starting a container that is running Postgres).

Expected behavior: the client should respect the configured query / connection timeouts and throw an error back to the caller

Known workarounds: surround each call to the client with `withTimeout()`, but it's not clear if there are any side-effects to this solution

```
runBlocking {
    (1..50).map {
        GlobalScope.launch {
            try {
                println("Before SELECT 1")
                connectionPool.sendQuery("SELECT 1").await()
                println("After SELECT 1")
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }.joinAll()
}
```

```
io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: localhost/127.0.0.1:15432
Caused by: java.net.ConnectException: finishConnect(..) failed: Connection refused
	at io.netty.channel.unix.Errors.throwConnectException(Errors.java:124)
	at io.netty.channel.unix.Socket.finishConnect(Socket.java:251)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.doFinishConnect(AbstractEpollChannel.java:673)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.finishConnect(AbstractEpollChannel.java:650)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.epollOutReady(AbstractEpollChannel.java:530)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:470)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
```

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -10,3 +10,11 @@ To run tests against a live mysql instance we are using [test-containers](https:
 Docker is required to run the tests.  
 No need to run any special script for a local mysql instance, however, to speed up the tests, you can run the `resources/run-docker-mysql.sh` script.  
 Note that all tables are created with `CREATE TEMPORARY TABLE`. In case you want to debug it might be helpful to remove the `TEMPORARY`.
+
+# Building locally
+
+`./gradlew clean build`
+
+To fix formatting:
+
+`./gradlew ktlintFormat`

--- a/pool-async/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
+++ b/pool-async/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
@@ -296,8 +296,8 @@ private class ObjectPoolActor<T : PooledObject>(
 
     private fun checkWaitingFuturesForTimeout() {
         val queryTimeout = configuration.queryTimeout
-            ?: // no timeout
-            return
+        // no timeout
+            ?: return
         while (!waitingQueue.isEmpty()) {
             val futureHolder = waitingQueue.peek()
             if (futureHolder.timeElapsed > queryTimeout) {

--- a/pool-async/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
+++ b/pool-async/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
@@ -125,7 +125,7 @@ internal constructor(
             return
         }
         logger.trace { "testAvailableItems - starting" }
-        val offered = actor.offer(TestAvailableItems())
+        val offered = actor.offer(TestPoolItems())
         if (!offered) {
             logger.warn { "failed to offer to actor - testAvailableItems()" }
         }
@@ -195,7 +195,7 @@ private class Created<T : PooledObject>(
     }
 }
 
-private class TestAvailableItems<T : PooledObject> : ActorObjectPoolMessage<T>()
+private class TestPoolItems<T : PooledObject> : ActorObjectPoolMessage<T>()
 private class Close<T : PooledObject>(val future: CompletableFuture<Unit>) :
     ActorObjectPoolMessage<T>()
 
@@ -207,8 +207,8 @@ private class ObjectPoolActor<T : PooledObject>(
     private val channelProvider: () -> SendChannel<ActorObjectPoolMessage<T>>
 ) {
 
-    private val availableItems: Queue<PoolObjectHolder<T>> = LinkedList<PoolObjectHolder<T>>()
-    private val waitingQueue: Queue<CompletableFuture<T>> = LinkedList<CompletableFuture<T>>()
+    private val availableItems: Queue<PoolObjectHolder<T>> = LinkedList()
+    private val waitingQueue: Queue<ObjectHolder<CompletableFuture<T>>> = LinkedList()
     private val inUseItems = WeakHashMap<T, ItemInUseHolder<T>>()
     private val inCreateItems = mutableMapOf<Int, ObjectHolder<CompletableFuture<out T>>>()
     private var createIndex = 0
@@ -216,7 +216,7 @@ private class ObjectPoolActor<T : PooledObject>(
 
     val availableItemsList: List<T> get() = availableItems.map { it.item }
     val usedItemsList: List<T> get() = inUseItems.keys.toList()
-    val waitingForItemList: List<CompletableFuture<T>> get() = waitingQueue.toList()
+    val waitingForItemList: List<CompletableFuture<T>> get() = waitingQueue.toList().map { it.item }
     val usedItemsSize: Int get() = inUseItems.size
     val waitingForItemSize: Int get() = waitingQueue.size
     val availableItemsSize: Int get() = availableItems.size
@@ -229,7 +229,7 @@ private class ObjectPoolActor<T : PooledObject>(
             is Take<T> -> handleTake(message)
             is GiveBack<T> -> handleGiveBack(message)
             is Created<T> -> handleCreated(message)
-            is TestAvailableItems<T> -> handleTestAvailableItems()
+            is TestPoolItems<T> -> handleTestPoolItems()
             is Close<T> -> handleClose(message)
             else -> XXX("no handle for message $message")
         }
@@ -240,8 +240,8 @@ private class ObjectPoolActor<T : PooledObject>(
         logger.trace { "scheduleNewItemsIfNeeded - $poolStatusString" }
         // deal with inconsistency in case we have items but also waiting futures
         while (availableItems.size > 0 && waitingQueue.isNotEmpty()) {
-            val future = waitingQueue.peek()
-            val wasBorrowed = borrowFirstAvailableItem(future)
+            val futureHolder = waitingQueue.peek()
+            val wasBorrowed = borrowFirstAvailableItem(futureHolder.item)
             if (wasBorrowed) {
                 waitingQueue.remove()
                 logger.trace { "scheduleNewItemsIfNeeded - borrowed object ; $poolStatusString" }
@@ -274,7 +274,7 @@ private class ObjectPoolActor<T : PooledObject>(
                 it.key.destroy()
             }
             inUseItems.clear()
-            waitingQueue.forEach { it.completeExceptionally(PoolAlreadyTerminatedException()) }
+            waitingQueue.forEach { it.item.completeExceptionally(PoolAlreadyTerminatedException()) }
             waitingQueue.clear()
             inCreateItems.values.forEach {
                 it.item.completeExceptionally(PoolAlreadyTerminatedException())
@@ -286,24 +286,44 @@ private class ObjectPoolActor<T : PooledObject>(
         }
     }
 
-    private fun handleTestAvailableItems() {
+    private fun handleTestPoolItems() {
         sendAvailableItemsToTest()
         checkItemsInCreationForTimeout()
         checkItemsInTestOrQueryForTimeout()
+        checkWaitingFuturesForTimeout()
         logger.trace { "testAvailableItems - done testing" }
+    }
+
+    private fun checkWaitingFuturesForTimeout() {
+        val queryTimeout = configuration.queryTimeout
+            ?: // no timeout
+            return
+        while (!waitingQueue.isEmpty()) {
+            val futureHolder = waitingQueue.peek()
+            if (futureHolder.timeElapsed > queryTimeout) {
+                logger.trace { "checkWaitingFuturesForTimeout - timeout waiting future after ${futureHolder.timeElapsed} ms" }
+                // should timeout future
+                waitingQueue.remove()
+                futureHolder.item.completeExceptionally(TimeoutException("timeout while waiting in queue after ${futureHolder.timeElapsed} ms"))
+            } else {
+                // we assume that once we found a future that should still wait
+                // all futures after him should also wait because they arrived later
+                return
+            }
+        }
     }
 
     private fun checkItemsInTestOrQueryForTimeout() {
         inUseItems.entries.removeAll { entry ->
             val holder = entry.value
             val item = entry.key
-            var timeouted = false
+            var itemWasTimeout = false
             if (holder.isInTest && holder.timeElapsed > configuration.testTimeout) {
                 logger.trace { "failed to test item ${item.id} after ${holder.timeElapsed} ms, will destroy it" }
                 holder.cleanedByPool = true
                 item.destroy()
                 holder.testFuture!!.completeExceptionally(TimeoutException("failed to test item ${item.id} after ${holder.timeElapsed} ms"))
-                timeouted = true
+                itemWasTimeout = true
             }
             if (!holder.isInTest && configuration.queryTimeout != null &&
                 holder.timeElapsed > configuration.queryTimeout + extraTimeForTimeoutCompletion
@@ -311,9 +331,9 @@ private class ObjectPoolActor<T : PooledObject>(
                 logger.error { "timeout query item ${item.id} after ${holder.timeElapsed} ms and was not cleaned by connection as it should, will destroy it - timeout is ${configuration.queryTimeout}" }
                 holder.cleanedByPool = true
                 item.destroy()
-                timeouted = true
+                itemWasTimeout = true
             }
-            timeouted
+            itemWasTimeout
         }
     }
 
@@ -438,7 +458,7 @@ private class ObjectPoolActor<T : PooledObject>(
                 logger.trace { "add ${message.returnedItem.id} to available items, size is ${availableItems.size}" }
             } else {
                 val waitingFuture = waitingQueue.remove()
-                message.returnedItem.borrowTo(waitingFuture, validate = false)
+                message.returnedItem.borrowTo(waitingFuture.item, validate = false)
             }
         } catch (e: Throwable) {
             logger.trace(e) { "GiveBack caught exception, so destroying item ${message.returnedItem.id} " }
@@ -491,7 +511,7 @@ private class ObjectPoolActor<T : PooledObject>(
                 createObject(message.future)
             } else {
                 if (waitingQueue.size < configuration.maxQueueSize) {
-                    waitingQueue.add(message.future)
+                    waitingQueue.add(ObjectHolder(message.future))
                     logger.trace { "no items available (${inUseItems.size} used), added to waiting queue (${waitingQueue.size} waiting)" }
                 } else {
                     logger.trace { "no items available (${inUseItems.size} used), and the waitQueue is full (${waitingQueue.size} waiting)" }
@@ -517,8 +537,7 @@ private class ObjectPoolActor<T : PooledObject>(
     }
 
     private fun validate(item: T) {
-        val tried = objectFactory.validate(item)
-        when (tried) {
+        when (val tried = objectFactory.validate(item)) {
             is Failure -> throw tried.exception
         }
     }

--- a/pool-async/src/test/java/com/github/jasync/sql/db/pool/ActorBasedObjectPoolTest.kt
+++ b/pool-async/src/test/java/com/github/jasync/sql/db/pool/ActorBasedObjectPoolTest.kt
@@ -68,6 +68,26 @@ class ActorBasedObjectPoolTest {
     }
 
     @Test
+    fun `take operation that is waiting in queue - when create is stuck should be timeout`() {
+        tested = ActorBasedObjectPool(
+            factory, configuration.copy(maxObjects = 1, queryTimeout = 10), false
+        )
+        // first item is canceled when the create fails
+        tested.take()
+        val takeSecondItem = tested.take()
+        factory.creationStuck = true
+        Thread.sleep(20)
+        // third item is not timeout immediately
+        val takeThirdItem = tested.take()
+        tested.testAvailableItems()
+        assertThat(takeThirdItem.isCompletedExceptionally).isFalse
+        await.untilCallTo { takeSecondItem.isCompletedExceptionally } matches { it == true }
+        Thread.sleep(20)
+        tested.testAvailableItems()
+        await.untilCallTo { takeThirdItem.isCompletedExceptionally } matches { it == true }
+    }
+
+    @Test
     fun `basic take operation - when create is little stuck should not be timeout (create timeout is 5 sec)`() {
         tested = ActorBasedObjectPool(
             factory, configuration.copy(createTimeout = 5000), false

--- a/pool-async/src/test/resources/logback.xml
+++ b/pool-async/src/test/resources/logback.xml
@@ -1,0 +1,20 @@
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder>
+            <pattern>[%level][%thread][%d][%c{5}] %msg%ex%n</pattern>
+        </encoder>
+    </appender>
+
+    <appender name="FILE" class="ch.qos.logback.core.FileAppender">
+        <File>target/mysql-async-tests.log</File>
+        <encoder>
+            <pattern>[%level][%thread][%d][%c{5}] %msg%ex%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="TRACE">
+        <appender-ref ref="STDOUT"/>
+        <appender-ref ref="FILE"/>
+    </root>
+</configuration>

